test/conftest.py

```diff
@@ -28,6 +31,77 @@ def pytest_addoption(parser: pytest.Parser) -> None:
     )
 
 
+def docker_warmup(request: pytest.FixtureRequest) -> None:
+    machine = request.config.getoption("--run-emulation", default=None)
+    if machine is None:
+        archs = [arch.value for arch in Architecture.auto_archs("linux")]
+    elif machine == "all":
+        archs = EMULATED_ARCHS
+    else:
+        archs = [machine]
+    options = Options(
+        platform="linux",
+        command_line_arguments=CommandLineArguments.defaults(),
+        env={},
+        defaults=True,
+    )
+    build_options = options.build_options(None)
+    assert build_options.manylinux_images is not None
+    assert build_options.musllinux_images is not None
+    images = [build_options.manylinux_images[arch] for arch in archs] + [
+        build_options.musllinux_images[arch] for arch in archs
+    ]
+    for image in images:
+        try:
+            command = (
+                "manylinux-interpreters ensure $(manylinux-interpreters list 2>/dev/null | grep -v graalpy) &&"
+                "cpython3.13 -m pip download -d /tmp setuptools wheel pytest"
+            )
+            container_id = subprocess.run(
+                ["docker", "create", image, "bash", "-c", command],
+                text=True,
+                check=True,
+                stdout=subprocess.PIPE,
+            ).stdout.strip()
+            subprocess.run(["docker", "start", container_id], check=True, stdout=subprocess.DEVNULL)
+            exit_code = subprocess.run(
+                ["docker", "wait", container_id], text=True, check=True, stdout=subprocess.PIPE
+            ).stdout.strip()
+            if exit_code == "0":
+                subprocess.run(
+                    ["docker", "commit", container_id, image], check=True, stdout=subprocess.DEVNULL
+                )
+            subprocess.run(["docker", "rm", container_id], check=True, stdout=subprocess.DEVNULL)
+        except subprocess.CalledProcessError:
+            print(f"failed to warm-up {image} image")
+
+
+@pytest.fixture(scope="session", autouse=True)
+def docker_warmup_fixture(
+    request: pytest.FixtureRequest, tmp_path_factory: pytest.TempPathFactory, worker_id: str
+) -> None:
+    # if we're in CI testing linux, let's warm-up docker images
+    if detect_ci_provider() is None or platform != "linux":
+        return None
+    if request.config.getoption("--run-emulation", default=None) is not None:
+        # emulation tests only run one test in CI, caching the image only slows down the test
+        return None
+
+    if worker_id == "master":
+        # not executing with multiple workers
+        return docker_warmup(request)
+
```

Comment:
Is this early return needed? I can't see the downside of the locking in the single-worker case?

Reply:
This was taken from https://pytest-xdist.readthedocs.io/en/stable/how-to.html#making-session-scoped-fixtures-execute-only-once

I asked myself the same question but did not add a comment.
I can't see the downside of the locking but I can see one in writing to `tmp_path_factory.getbasetemp().parent` which might not be so temporary anymore in single worker mode.